### PR TITLE
Add Render Sciname function for formatting taxonomy

### PR DESCRIPTION
**Pull request by MuchQuak**
# Issue #https://github.com/BioKIC/symbiota-docs/issues/679

# Summary
There are other places that do a similar job, however none of them just take a string and render it without much of a fuss.

This function is not perfect but it will beat just rendering everything in italics which is how this is mostly handled. Notably it doesn't handle tradenames as I could not find much info of parsing indicators for tradenames. If they are always tied to genus + species it would be possible to infer them.

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [x] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [x] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [x] Comment which GitHub issue(s), if any does this PR address
- [x] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

### CODE REVIEW

File: classes/utilities/TaxonomyUtil.php [hunk 1]
```diff
@@ -348,6 +349,46 @@ public static function buildHierarchyEnumTree($conn = null, $taxAuthId = 1){
...
+
+		$html = '';
+
+		for($i=0; $i < count($matches); $i++) {
+			$part = $matches[$i];
+			$previousPart = $i != 0? $matches[$i - 1]: false;
+
+			if(in_array($part, ['X', 'x', '×', '&times;'])) {
+				$html .= '&times; ';
+			} else if($part === 'subsp.' || $part === 'var.' || strlen($part) > 0 && $part[0] === "'") {
```
Comment (Atticus29) [inline classes/utilities/TaxonomyUtil.php:379]: We can also have "ssp.", "f.", "fo.".

File: classes/utilities/TaxonomyUtil.php [hunk 2]
```diff
@@ -348,6 +349,46 @@ public static function buildHierarchyEnumTree($conn = null, $taxAuthId = 1){
...
+	 * Tradenames will default display to italics. This is not ideal, but need
+	 * and way of identifing them consistently.
+	 *
+	 * Can parse hybrid strings using 'X', 'x', '×', or '&times;' and resolve
+	 * cultivars that use " or ' as denotations
+	 *
+	 * @param Type $scientificName
+	 * @return string
+	 **/
+	function renderSciname(string $scientificName): string {
```
Comment (Atticus29) [inline classes/utilities/TaxonomyUtil.php]: This should probably be a public static function, right? I was also trying to put a lot of the sciName processing in the Trait file itself (traits/TaxonomyTrait.php), but this feels just as good a place.

**Review by Atticus29 — CHANGES_REQUESTED** [review]
Generally looks good to me. Unfortunately, I think it's common to place the author in front of the cultivar epithet and tradename, and it's also convention to not italicize the trade name, as you noted in the description.

I think that implementing this should probably not happen until the above issues are resolved. Otherwise, we're swapping one problem for another that people seem to care more about.

I'm happy to approve this once the comments are addressed. But implementation should be low priority IMHO.

File: classes/utilities/TaxonomyUtil.php (hunk 1 above)
Comment (MuchQuak) [reply classes/utilities/TaxonomyUtil.php:379]: What do the "ssp.", "f.", "fo." stand for?

File: classes/utilities/TaxonomyUtil.php (hunk 2 above)
Comment (MuchQuak) [reply classes/utilities/TaxonomyUtil.php]: This makes sense, not sure why it wasn't in the first place

Comment (MuchQuak) [reply classes/utilities/TaxonomyUtil.php]: Also I think I trying to keep static function out of traits for sake for clarity as it would be kind of bizarre to call a static function on a class that had that function from a trait.

File: classes/utilities/TaxonomyUtil.php (hunk 1 above)
Comment (MuchQuak) [reply classes/utilities/TaxonomyUtil.php:379]: I should note in multiple places the entire sci name is just italicized so I imagine anything would be better. Also if we could compile some running test data of edge cases it would help a lot. I am pretty familiar with how the hybrids are handled so I got plenty of test data on that front but not some much with trade names and authors in front of cultivar.

Comment (Atticus29) [reply classes/utilities/TaxonomyUtil.php:379]: ssp. = subspecies
f. and fo. = forma (apparently a taxon rank below even variety.

When it gets decided whether we want a shared 3.3 db on the dev servers, I'll start creating some edge case taxa.

**Review by Atticus29 — APPROVED** [review]
I think that this was the only essential fix. Thanks!

**Comment by MuchQuak** [comment]
> I think that this was the only essential fix. Thanks!

Sweet, I will add forma and ssp. fo. and f. extension support before merging for 3.5 stuff

File: classes/utilities/TaxonomyUtil.php (hunk 1 above)
Comment (egbot) [reply classes/utilities/TaxonomyUtil.php:379]: Ideally, there shouldn't be any ssp. in the database because they get converted to subsp. upon import, but the data entry form does allow freeform entry of infraspecific rank designations, thus it's possible. Same with fo. vs f.

**Review by egbot — APPROVED** [review]
LGTM, and we can refine as we go forward, if needed.
